### ISSUE TITLE
Lint categorization

### ISSUE THREAD

**Issue by not-my-profile** [charliermarsh/ruff#1774]
I think there are two perspectives ruff users may have:

* using ruff as a replacement for an existing linter (such as pylint or flake8)
* using ruff from the get-go

The README currently groups rules by their origin. Which I think is suboptimal for both cases. For the first case our README only lists the lints implemented by ruff but doesn't tell you which (or how many) lints ruff is missing. For the second case you don't actually care about where the lints are coming from, you just want to see them grouped by topic.

So I think we should steal a nice feature from Clippy which are [lint categories](https://github.com/rust-lang/rust-clippy#readme). Clippy has the following categories:

* correctness: code that is outright wrong or useless
* suspicious: code that is most likely wrong or useless
* style: code that should be written in a more idiomatic way
* complexity: code that does something simple but in a complex way
* perf: code that can be written to run faster
* pedantic: lints which are rather strict or have occasional false positives
* nursery: new lints that are still under development
* cargo: lints for the cargo manifest

We could simply use the same categorization (except with cargo instead being pyproject as per #1472). I think we would however need some additional categories since there are several lint categories that Rust does not need, for example:

* syntax: code that is syntactically invalid (syntax errors cannot go by unnoticed in Rust)
* deprecated: functions that have been deprecated (Rust has a `#[deprecated]` macro for that)

What do you think about this?

**Comment by charliermarsh** [comment]
I agree that we should kick off a re-categorization effort. Like #1773, will reply in a bit with some more thoughts :)

**Comment by charliermarsh** [comment]
I generally agree. I need to give some thought to the right categorization since this will be hard to undo.

My gut reaction is that... we may need more categories? Two that come to mind for me are `docstrings` and `typing` (or `type-annotations`, or whatever).

> For the first case our README only lists the lints implemented by ruff but doesn't tell you which (or how many) lints ruff is missing.

It's not really relevant for this debate, but: in the majority of cases, we support the entire plugin. There are a few exceptions.

**Comment by sbrugman** [comment]
Overview of [categories](https://github.com/dosisod/refurb/blob/master/docs/categories.md) used in `refurb`

Plugin ticket: https://github.com/charliermarsh/ruff/issues/1348

**Comment by tgross35** [comment]
+1 for simplifying categories, it's kind of a mess with _60 categories_ now (wow that's a lot).

I think it is totally reasonable to recategorize everything and assign coherent numbers for the error output (could use the `R` prefix and not just `RUF` since I don't see that used anywhere). With https://github.com/astral-sh/ruff/issues/1256, it would be useful to assign different levels to each lint group and allow the user to override the options, e.g.:

```toml
[lints]
suspicious = "allow"
docstrings = "error"
```

I _do_ think it is still important to keep the `flake8` numbers/groups around in the docs and usable for configuration - just since people are going to be looking for flake8 compatibility for the next decade. Doesn't need to be the main promoted style, but I have to imagine it would slow new adoption if `enable = ["E", "W"]` doesn't continue to work. Maybe there could be a "fix" tool that upgrades the flake8 style codes to whatever the new style is?

**Comment by charliermarsh** [comment]
Yeah, that's roughly my thinking too: recategorize the rules, but try to preserve support for the Flake8 compatibility layer at runtime + ship tooling to automatically migration an existing configuration file.

**Comment by MichaReiser** [comment]
Related requests are to disable all formatter rules, or deprecated rules

https://github.com/astral-sh/ruff/issues/10225
https://github.com/astral-sh/ruff/issues/9778

**Comment by kaddkaka** [comment]
As another reference, the c++ checker clang-tidy (https://clang.llvm.org/extra/clang-tidy/checks/list.html) has amongst others, these categories:

- `performance`
- `modernize`
- `readability`
- `concurrency`
- `bugprone`

**Comment by Goldziher** [comment]
this is now a convention adopted by BiomeJS and OXC.

**Comment by oXidFoX** [comment]
My use case would be to check if people have already run the ruff autofix on their code.
I would like to have the "all-autofixable" subset of rules for use with the command `ruff check --select ALL-autofixable`.

This subset would be the same as the one used by `ruff check --fix`, including the extends/ignores and the unsafe rules, if selected, but for a simple `ruff check`.

Example:
```shell
# auto fix everything it can
ruff check --fix --select ALL --ignore D203
# should return nothing more or less than the previous command
ruff check --select ALL-autofixable --ignore D203
```

Would this be possible?

**Comment by allanlewis** [comment]
> I would like to have the "all-autofixable" subset of rules for use with the command `ruff check --select ALL-autofixable`.

Is this not what `--fix-only` does?

**Comment by oXidFoX** [comment]
> > I would like to have the "all-autofixable" subset of rules for use with the command `ruff check --select ALL-autofixable`.
> 
> Is this not what `--fix-only` does?

no, `--fix-only` does the fix, without reporting. I would like the "check only" version.

[As the documentation states](https://docs.astral.sh/ruff/settings/#fix-only): _Apply fixes to resolve lint violations, but don't report on, or exit non-zero for, leftover violations. Implies `--fix`._

**Comment by allanlewis** [comment]
> no, `--fix-only` does the fix, without reporting. I would like the "check only" version.

I see - so more like "is anything fixable".

**Comment by oXidFoX** [comment]
> > no, `--fix-only` does the fix, without reporting. I would like the "check only" version.
> 
> I see - so more like "is anything fixable".

exactly this !

**Comment by thernstig** [comment]
https://github.com/astral-sh/ruff/issues/24979 was closed in favor of this, so adding this as reference:

> https://docs.astral.sh/ruff/formatter/#conflicting-lint-rules mentions several lint rules that are conflicting with the formatter, when both are used.
>
> Instead of users having to keep track of these conflicting rules, and updating them whenever new additions are made to that table, it would be nice to have a new ruleset such as CONFLICT that disables all of them. If the ruleset is maintained and updated when new rules are found to be conflicting, users do not have to act to keep track of that list, but a simple upgrade of Ruff would solve it as the new rules would be brought in by the ruleset.

**Comment by flying-sheep** [comment]
As @ntBre suggested in https://github.com/astral-sh/ruff/issues/12491#issuecomment-4770999674, the new [RUF076](https://docs.astral.sh/ruff/rules/pytest-fixture-autouse/) is a great candidate for a “very opinionated” category

**Comment by flying-sheep** [comment]
We should add the ability to filter rules on the website. E.g. I want to see all “pedantic” rules to know if I should disable the category and enable things individually or the other way around.

**Comment by ntBre** [comment]
Agreed, I have a couple of draft PRs up exploring docs improvements for the categories: https://github.com/astral-sh/ruff/pull/28087 and https://github.com/astral-sh/ruff/pull/28099. We're still working out the best way to represent this, though.

For `pedantic` specifically, our recommendation (also part of #28087), is to disable the category by default and pick out individual rules you're interested in.

The category information is also available in the JSON output if you want to hack around the missing docs in the meantime:

```console
❯ uvx ruff@0.16.5 rule --all --output-format=json |
  jq '.[] | select(.category == "pedantic") | .name' |
  head
"airflow3-removal"
"airflow3-moved-to-provider"
"airflow3-incompatible-function-signature"
"airflow3-suggested-update"
"airflow3-suggested-to-move-to-provider"
"airflow31-moved"
"commented-out-code"
"missing-type-function-argument"
"missing-type-args"
"missing-type-kwargs"
```
